[clang][OpenMP] Add 'allocator' modifier for 'allocate' clause.

ddpagan opened the pull request:
The 'allocator' modifier is now accepted in the 'allocate' clause. Added LIT tests covering codegen, PCH, template handling, and serialization for 'allocator' modifier.

Added support for allocator-modifier to release notes.

Testing
- New allocate modifier LIT tests.
- OpenMP LIT tests.
- check-all
- relevant sollve_vv test cases tests/5.2/scope/test_scope_allocate_construct.c

ddpagan committed:
[clang][OpenMP] Add 'allocator' modifier for 'allocate' clause.

The 'allocator' modifier is now accepted in the 'allocate' clause.
Added LIT tests covering codegen, PCH, template handling, and
serialization for 'allocator' modifier.

Added support for allocator-modifier to release notes.

Testing
- New allocate modifier LIT tests.
- OpenMP LIT tests.
- check-all
- relevant sollve_vv test cases
    tests/5.2/scope/test_scope_allocate_construct.c

ddpagan commented:
NOTE: OpenMP Support doc will be updated as well now that the PR # is available.

alexey-bataev commented on clang/lib/AST/OpenMPClause.cpp:
```diff
@@ -2242,9 +2247,16 @@ void OMPClausePrinter::VisitOMPAllocateClause(OMPAllocateClause *Node) {
+      OS << "(";
+      Allocator->printPretty(OS, nullptr, Policy, 0);
+      OS << ")";
+    } else
+      Allocator->printPretty(OS, nullptr, Policy, 0);
```
```suggestion
    } else {
      Allocator->printPretty(OS, nullptr, Policy, 0);
    }
```

ddpagan replied:
Fixed as suggested.

alexey-bataev commented on clang/lib/Parse/ParseOpenMP.cpp:
```diff
@@ -4539,6 +4539,8 @@ bool Parser::ParseOpenMPVarList(OpenMPDirectiveKind DKind,
   bool NeedRParenForLinear = false;
   BalancedDelimiterTracker LinearT(*this, tok::l_paren,
                                    tok::annot_pragma_openmp_end);
+  BalancedDelimiterTracker AllocateT(*this, tok::l_paren,
+                                     tok::annot_pragma_openmp_end);
```
Move it exactly where it used (lines 4807-4819)

ddpagan replied:
Fixed as suggested.

ddpagan committed:
Updates for review comments:
- Added '{' and '}' to else with single statement per coding standards.
- BalancedDelimiterTracker() definition moved to where it's used.

alexey-bataev reviewed (approved):
LG with a nit. Update OpenMPSupport.rst

ddpagan commented:
@alexey-bataev - I'll make that change in the next PR. Thanks for the review.
